Use SQLite to store torrents and fastresumes

sledgehammer999 opened the pull request:
I recently played with SQLite and I was fascinated by it. Then I realized how easy things could become for us if we used to save stuff.

Pros:
1. Far easier code for sorting torrents by queue during startup
2. Possibly much faster startup for users with many torrents because now we don't have to read multitudes of **small** files from disk
3. SQLite should have a very robust anti-corruption systems. Far better than anything we can engineer ourselves

I plugged in the new functionality by keeping the current design as much as possible. I didn't attempt to engineer a better code architecture around the new system.
Most of the code you see is old code moved around.

PS: Is anyone more knowledgeable with SQL? I wonder if we should split the table into 2 for performance reasons. One saving hash+metadata and one saving hash+fastresume+queue.

sledgehammer999 commented:
Special shoutout to [DB Browser for SQLite](https://sqlitebrowser.org/). It is a great lightweight that lets you inspect/modify/administer and SQLite db. Very useful for debugging.

glassez commented:
@sledgehammer999, that's fine! 
I'm glad my old idea finally sprung. This will undoubtedly be a step forward. Although there is still work to be done, starting from what and how we should store in the tables, and ending with the details of the implementation. I'll start commenting on specific things later.

Chocobo1 commented on src/app/app.pri:
```diff
@@ -40,3 +40,4 @@ stacktrace {
 
 # upgrade code
 HEADERS += $$PWD/upgrade.h
+SOURCES += $$PWD/upgrade.cpp
```
a bit unrelated to this PR, how about move this to the `SOURCES` above (and also `HEADERS`).

sledgehammer999 replied:
Sure.
I don't know the initial reason for this. Maybe it was here because "upgrade" was a temporary solution. But now it pretty much is a stable part.

Chocobo1 commented on src/base/bittorrent/torrenthandle.h:
```diff
@@ -346,7 +346,7 @@ namespace BitTorrent
 #endif
         void renameFile(int index, const QString &name);
-        bool saveTorrentFile(const QString &path);
+        QByteArray metadata();
```
`QByteArray metadata(); const`

Chocobo1 commented on src/base/bittorrent/private/resumedatasavingmanager.cpp:
```diff
@@ -28,34 +28,47 @@
 
-    Utils::Fs::forceRemove(filepath);
+    if (!query.exec())
+        Logger::instance()->addMessage(QString("Couldn't save data in '%1'. Error: %2")
```
`tr("ResumeDataSavingManager couldn't save data in '%1'. Error: %2")`

sledgehammer999 commented on src/base/utils/db.h:
```diff
@@ -0,0 +1,44 @@
+        const char COL_METADATA[] = "metadata";
+        const char COL_FASTRESUME[] = "fastresume";
+        const char COL_QUEUE[] = "queue";
+        const char COL_VERSION[] = "version";
```
I prefer kepping this in logical order. In the order we use them in the DB.

Chocobo1 replied:
>In the order we use them in the DB.

Do you mean there is some order between COL_* fields?

sledgehammer999 replied:
>Do you mean there is some order between COL_* fields?

I mean the logical order we use when issuing the `CREATE TABLE` statement. For example, we logically think that the "hash" column is the first one when thinking about each row.

glassez commented on src/base/utils/db.h:
```diff
@@ -0,0 +1,44 @@
+
+#pragma once
+
+namespace Utils
```
There are no database related utilities in this file. It's just database configuration for some particular application component (BitTorrent). So it should be under "base/bittorrent".

glassez commented:
>I plugged in the new functionality by keeping the current design as much as possible.

Too much.
Since we stored some additional data along with the "fastresume" data, we had to request the creation of "fastresume" data each time we had anything changed (e.g. save path, torrent name, etc.), then we injecte these additional fields in "fastresume" data, encode it and write to file.
We don't have to do it that way anymore. We can store them in separate columns and update them independently.
Even more. If we look at the libtorrent "fastresume" data, we can see that most of its fields has the behavior described above. E.g. when torrent is paused or resumed we receive an appropriate alert so we can just update "paused" and "auto_managed" columns and so on.
By and large, we need to request "fastresume" data generation just to store the current progress.
This may seem somewhat inconvenient due to libtorrent drawback (it requires "fastresume" data to be bencoded when we pass it to add_torrent_params). But this is slightly improved in the libtorrent-1.2 (now we just have to fill in the appropriate add_torrent_params fields).

sledgehammer999 commented:
I am all for decoupling our data from libtorrent's fastresume.
But this will need investigation on when to save our data and what triggers it. Aka reacting more closely to the triggering event.
This is work for the weekend though.

glassez commented:
>But this will need investigation on when to save our data and what triggers it.

At least all current "fastresume" saving events that are not fired by timer.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -535,7 +538,8 @@ Session::Session(QObject *parent)
     m_ioThread = new QThread(this);
-    m_resumeDataSavingManager = new ResumeDataSavingManager {m_resumeFolderPath};
+    const QDir resumeDataDir(m_resumeFolderPath);
+    m_resumeDataSavingManager = new ResumeDataSavingManager{resumeDataDir.absoluteFilePath(QLatin1String {Utils::DB::FILENAME})};
```
Do you really want to have database in BT_backup folder? Why? Not only it has the inappropriate name (we kept it for compatibility purposes only), a separate folder is not needed at all now.

sledgehammer999 replied:
A reason I did it is because we have the lock file there which kinda guarantees that we have permission to read/write in that folder. Which in turn will make exporting the torrents back to the all system more manageable.
I open to suggestions for a new location. Also for a better name for the DB/table.

LordNyriox commented on src/base/bittorrent/session.cpp:
```diff
@@ -2354,22 +2355,34 @@ void Session::exportTorrentFile(TorrentHandle *const torrent, TorrentExportFolde
              ((folder == TorrentExportFolder::Finished) && !finishedTorrentExportDirectory().isEmpty()));
 
+    using namespace Utils::DB;
+    QSqlQuery query;
```
@sledgehammer999, @glassez:  

Perhaps you should make a new class to manage SQLite read/write operations? 

Sort of like `SettingsStorage`?

sledgehammer999 replied:
@LordNyriox I'll have to see fist, but I don't think the DB code (due to the nature of SQL statements) can be abstracted away in to a class.
Any abstraction will be done last, because we need to agree on a format of storing data in the db.

sledgehammer999 commented:
Pushed new version of commits with these changes:
1. db.h moved under base/bittorrent
2. Namespace from `Utils::DB` to `DB`
3. Fixed issue where deleted torrents weren't removed from the DB due to missing variable escape

As for the other issues I'll comment inline.

sledgehammer999 commented on src/app/upgrade.cpp:
```diff
@@ -0,0 +1,351 @@
+
+        // Table doesn't exist. Probably 1st run.
+        // Input PRAGMAs and create table
+        query.exec(QLatin1String("PRAGMA auto_vacuum = FULL"));
```
Yes. It shrinks the file small when removing torrents. Why should we use it?

glassez replied:
>It shrinks the file small when removing torrents. 

Is that really necessary? The empty pages can be reused further. Additionally the doc says:
```
However, using auto_vacuum can lead to extra database file fragmentation. And auto_vacuum does not compact partially filled pages of the database
```
Maybe it's better to use VACUUM command from time to time or allow user to perform it?

sledgehammer999 commented on src/app/upgrade.cpp:
```diff
@@ -0,0 +1,351 @@
+        namespace libt = libtorrent;
+
+        libt::error_code ec;
+#if LIBTORRENT_VERSION_NUM < 10100
```
This is moved around code. I am going to remove it all in a fell swoop in new PR dropping all RC_1_0 code.

glassez replied:
"dropping all RC_1_0 code" is much easier job than current PR so I assume it will be done earlier and you still need to remove it here separately.

sledgehammer999 commented on src/app/upgrade.cpp:
```diff
@@ -0,0 +1,351 @@
+        libt::bencode(std::back_inserter(preparedResumeData), resumeData);
+    }
+
+    // This should be moved back into Session::initResumeFolder()
```
We need to do this check before trying to do any kind of upgrade. The upgrade code assumes that the db exists and is accessible and that the table exists in it.
The upgrade code is called before the Session is constructed.

glassez replied:
I'm surprised at this answer. I even thought for a moment that I was debating with another person...
>The upgrade code assumes that the db exists and is accessible and that the table exists in it.

This does not contradict the fact that this code should be in a different place.
>The upgrade code is called before the Session is constructed.

It's incorrect. At least it's done incorrectly. You ripped main Session code as well.

Main Session logic should be the following:
1. If database doesn't exist create it
2. Load torrents from database and go ahead.

When you add upgrade logic to it:
1. If database doesn't exist create it
2. If older version was used before this run try to import torrents from old "saving system"
3. Load torrents from database and go ahead.

All this logic should be done in Session code!

One thing that should be done before the Session is created is asking the user for accept the upgrade. To do this, we do not need to know anything about the existence of the database. All you need to know is whether the old qBittorrent version was used just before the current application run.

My main idea is "not to break the basic logic of the application for the sake of some additional".

sledgehammer999 commented on src/app/upgrade.cpp:
```diff
@@ -0,0 +1,351 @@
+    const QString backupFolderPath = Utils::Fs::expandPathAbs(specialFolderLocation(SpecialFolder::Data) + "BT_backup");
+    const QDir backupFolderDir(backupFolderPath);
+    const QString dbPath = backupFolderDir.absoluteFilePath(QLatin1String{FILENAME});
+    // If db exists don't try to upgrade
```
I think you misunderstood it here.
At the current moment this tries to updrade from the non-db way to the db-way. So if a db exists we don't try to migrate older files (.torrents+.fastresumes). It is assumed that the upgrade was done already and the user dropped new files at a later date.
Unless you suggest that we should always input into the db any kind of .torrent/.fastresume we find during startup.

sledgehammer999 replied:
Also in case you missed it: Each row has a `version` column. During startup if the version number is different from the supported one then that row isn't loaded. New versions will upgrade smaller version, while refusing to load bigger version.

glassez replied:
>I think you misunderstood it here.

Is it? Here's your code: 
```c++
                    // If db exists don't try to upgrade
		    if (QFile::exists(dbPath)) {
		        initiliazeDB(dbPath);
		        return true;
		    }
		    else {
		        if (ask && !userAcceptsUpgrade()) return false;
		    }
```

You ask the user for upgrade if the database doesn't exist. But it is incorrect if the database does not exist due to the fact that the user has never used qBittorrent (on this system).

I've suggested correct logic in another comment.
>Now, I realize that his approach has a drawback. 

Yes, it is.
>Should we instead create a 2nd table named "version" with a single col+row that will hold the version of the whole db?

I think "overall application versioning" I suggested in other comment should fit us. There is no need to have separate versions for each thing.

sledgehammer999 commented on src/app/upgrade.cpp:
```diff
@@ -0,0 +1,351 @@
+    initiliazeDB(dbPath);
+
+    QSqlDatabase db = QSqlDatabase::database();
+    db.transaction();
```
Sure. Truth is I was in a hurry to make everything work and forgot to come back to this later.

sledgehammer999 commented:
@glassez I pushed a few new commits. I didn't merge them back into the others in hope it is easier to review the changes.

I need your opinion. The last commit is unfinished and still WIP. I didn't have more time to tidy it up.
Can you take a look at it and consider the following:
I assume you agree to save each old fastresume  field as a table column, right? As you can see in the last commit, I save each setting as soon as it is changed. Here's where I want your input. Should the saving be done by the TorrentHandle class (like its done in my commit), or done by the Session class?
For example, when the torrent changes name, should the saving be done inside `TorrentHandle::setName()` or inside `Session::handleTorrentNameChanged()`?

I assume your answer will be that the Session should do that. It's more logical.
However, there are some settings which aren't publicly exposed by the TorrentHandle class. Like the `m_name` variable(`TorrentHandle::name()` doesn't return it verbatim). Is it ok if I extend the various Session::handleTorrent* methods to receive that argument directly from the caller?

In any case, as this last commit is very WIP, I have nothing solid. I merely was exploring the possibilities. So if you want to propose a better approach let me hear it (with a pseudo-example if possible).

WolfganP commented:
As I commented in https://github.com/qbittorrent/qBittorrent/issues/10115, I'm optimizing my RasPi setup to avoid using the SDcard or other slow storage units.

I tried to follow up this PR changes to check where the new database file will be stored, it's need for persistence across boots and the interaction with the portable settings; but I was unable to.

@sledgehammer999  do you mind to comment on it?

glassez commented:
>I assume you agree to save each old fastresume field as a table column, right?

Right.
But I propose to store ALL fastresume data in separate columns.

>Should the saving be done by the TorrentHandle class (like its done in my commit), or done by the Session class?

Although some logic is currently delegated to TorrentHandle class, I would prefer to have it centrally in the Session class and leave TorrentHandle only as a convenient interface.
>Is it ok if I extend the various Session::handleTorrent* methods to receive that argument directly from the caller?

I don't mind doing it that way to begin with. There may be a better way later.
>So if you want to propose a better approach let me hear it (with a pseudo-example if possible).

Ok.

sledgehammer999 commented:
>But I propose to store ALL fastresume data in separate columns.

You mean each field(even libtorrent native) in the fastresume in a separate column?
But what do we want to accomplish with that?
Libtorrent will still generate them as a whole anyway. We will need to manually parse them into separate columns. Then manually combine them into a fastresume to feed into libtorrent. Also we would need to always keep our parsers/tables updated with each revision of the fastresume spec.

glassez commented on src/base/bittorrent/db.h:
```diff
@@ -0,0 +1,59 @@
+    const char COL_HAS_FIRST_LAST_PIECE_PRIO[] = "hasFirstLastPiecePrio";
+    const char COL_SEQUENTIAL[] = "sequential";
+
+    const char TABLE_VERSION_NAME[] = "version";
```
If you still want to have version for particular database it's better to use "user_version" pragma.

WolfganP commented:
Just for the sake of documentation and other approaches to manage tmpfs needed folders (ie /var/log/*), interesting discussion at https://community.openenergymonitor.org/t/better-handling-of-files-in-tmpfs-var-log/9315

glassez commented:
>You mean each field(even libtorrent native) of the fastresume in a separate column?

Yes.
I do not insist on it yet, but just want to evaluate all the pros and cons. We need to do this now as accurately as possible to avoid multiple database schema changes in the future.
>But what do we want to accomplish with that?

The same things as for custom resume data members.
>Libtorrent will still generate them as a whole anyway. 

Yes. But in many cases we can avoid all data generation when we need to update only one field. I.e. when some even occurs we receive an appropriate alert so we can update just one (or several) field instead of asking full data from libtorrent.
>We will need to manually parse them into separate columns.

You're wrong. On the contrary, we avoid encoding the data in bencoded forms.
We get "fastresume" data either as dictionary in libtorrent 1.1 or as add_torrent_params in libtorrent 1.2 so it can be easily stored as separate columns.
>Then manually combine them into a fastresume to feed into libtorrent.

Yes. But it's not so big problem. Besides, it only applies to libtorrent 1.1. libtorrent 1.2 accepts all resume data as separate fields of add_torrent_params struct.

Additionally storing data in separate columns should make it easier to manage database and debug the application.

glassez commented on src/base/bittorrent/db.h:
```diff
@@ -0,0 +1,59 @@
+    const char MAIN_CONNECTION_NAME[] = "torrentsMain";
+    const char FILENAME[] = "torrentData.db";
+
+    const char TABLE_NAME[] = "torrents";
```
There are many discussions on this topic. Personally, I prefer to use singular for table names (e.g. "torrent"). Since SQL is a future of qBittorrent, let's make a project-wide rule about it.

CyrusNajmabadi commented:
I haven't looked through this entire PR, but as someone recommending this change I thought i would throw in a little bit of info that I learned from making the same change in https://github.com/dotnet/roslyn.

Specifically, if your app is write-heavy (which Roslyn is), then you'll very likely benefit hugely from setting `journal_mode=wal` and `synchronous=normal`.  You can see where we set that ourselves here: https://github.com/dotnet/roslyn/blob/b03098e236c958602783a929f55e355328e49518/src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage.cs#L224-L235

Documentation on these values are here:
https://sqlite.org/pragma.html#pragma_synchronous
https://www.sqlite.org/wal.html#performance_considerations

Importantly, moving to WAL helps heavy-write performance a **Ton**.  'normal' mode also helps quite a bit, while not introducing concerns about database corruption.  i.e. your database will always be corruption-free, and will always only contain the results of some previously completed transaction.  However, you risk an exceptionally tiny change that *some committed transaction* **might** be lost if you lose power at some precise moment.  Again, your data will not be corrupt. Simply, you might lose a transaction that in 'full' mode would reapply if power is lost at a specific moment.

For the Roslyn use case, this was perfect.  A lost transaction simply meant we'd recompute the data later on.  For QB, it's unclear what it would mean.  I get the sense that you ahve to validate data anyways on a non-clean shutdown.  So i imagine things will be fine for you.

Just something to consider in order to ensure that your DB is not a bottleneck if you are writing data to it non-stop.

--

Note: as someone who has a huge amount of torrents, I have a personal vested interest here.  I dont' know what the write patterns are for QB.  But, for example, if it routinely making thousands of transactions because there are all these torrents, then WAL will go a long way to making that crawl, to making it something you'd never notice.

Thanks!

CyrusNajmabadi commented on src/base/bittorrent/session.cpp:
```diff
@@ -1949,13 +1916,11 @@ bool Session::deleteTorrent(const QString &hash, bool deleteLocalFiles)
+    using namespace DB;
+    QSqlDatabase db = QSqlDatabase::database(MAIN_CONNECTION_NAME);
+    QSqlQuery query(db);
+    query.exec(QString("DELETE FROM %1 WHERE %2 = '%3'").arg(TABLE_NAME, COL_HASH, torrent->hash()));
```
are you constantly executing these queries?  if so, it's good to cache them, and then update the args using hte cache query in the future.  it saves expensive costs in sqlite parsing and evaluating these strings. (note: only important if this is done a lot).

CyrusNajmabadi replied:
I was talking in a general sense. Not about this query in particular.  For example, if you are constantly quering the DB, it can help significantly to cache the query.  

For example, say a user has many thousand torrents and you're going to execute the same query back to back for each torrent, then this can help quite a bit.

I agree it doesn't need to be done for something that happens rarely

glassez replied:
>I was talking in a general sense. Not about this query in particular.

Sorry. I just saw this suggestion having a general meaning in your other comment. Here it was attached to a specific place of the code, so it looks like it relates to this particular case.

Having prepared statement is a good idea in case of frequently invoked query.

CyrusNajmabadi commented on src/base/bittorrent/session.cpp:
```diff
@@ -3691,12 +3663,22 @@ void Session::handleTorrentMetadataReceived(TorrentHandle *const torrent)
+        query.prepare(QString("UPDATE %1 "
+                              "SET %2 =  :metadata "
+                              "WHERE %3 = :hash")
+                      .arg(TABLE_NAME, COL_METADATA, COL_HASH));
```
my other recommendation would be to encapsulate this behind some class that just exposes the operations you want to perform.  u.e. "updateHash".  That way you can change things behind the scenes, without having to update all these places.  It also isolates you from leaking out the details of tehse tables/columns and whatnot to all this consumptive code.

sledgehammer999 commented:
Rebased PR on top of current master, and addressed the namespace comment.
I'll continue work tomorrow.

sledgehammer999 commented:
@glassez I finished reworking the code in order to save our custom fields as separate columns in the DB.
I worked this code in fragments in the course of many days. So I might have missed something.
Look at the general picture. If the direction is ok then I can proceed to add error handling, squash commits, update the 2 other commits so you can do a proper review.

sledgehammer999 commented:
>Has any testing been done to validate no performance issues when there are thousands of torrents? Sqlite can def have perf issues depending on certain usage patterns.

Not yet. We need to solidify the table/db format first.
Then I'll look into ways on how to wrap the bulk fastresume writes in sql transactions. This should minimize performance degradation.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -3588,6 +3686,21 @@ const CacheStatus &Session::cacheStatus() const
+        if (ver != BitTorrent::DB::VERSION) {
+            LogMsg(tr("Can't load torrents from DB. DB version: %1. Program version: %2").arg(ver).arg(BitTorrent::DB::VERSION),
+                   Log::WARNING);
+            return;
```
Backward compatibility must be ensured in some way, otherwise the user will not be able to update the application without the risk of losing their torrents or stopping the application. But this is only a matter of implementation, because we know exactly the structure of the database of previous versions.
However, forward compatibility is generally an unattainable goal, so we must still limit the frequency of database format changes. We need to think about it as carefully as possible.

I had a crazy idea how to provide the possibility of database downgrade that provide some kind of forward compatibility (but "paranoid" will strongly against due to possible security issues). It lies in the fact that we could put SQL queries that perform a database downgrade in some special database table. Then, if a newer database version is found by application, it could retrieve the corresponding SQL queries and execute them.

glassez commented:
@sledgehammer999, having an concept of the target design of the feature, I could even help with some code snippets. Otherwise, I can only offer some more or less trivial helpers.

glassez commented:
@sledgehammer999, at this stage, I found one fundamental error.
The only reliable way to have a consistent torrent queue is to write it all at once. I fixed it recently, but here you degrade again, writing torrent positions separately.

sledgehammer999 commented:
@glassez I have tidied up the commits. Error handling is missing though. IMO, you can have a better idea now.

>What data and in what format are you going to store?

A table that stores the hash, metadata, fastresume(bendoced), queue and the rest of the custom `qBt-` fields per row.
The table definition looks like this
```c++
query.exec(QString("CREATE TABLE %1 "
                           "("
                           "%2 TEXT PRIMARY KEY NOT NULL, " // hash
                           "%3 BLOB NOT NULL, " // metadata
                           "%4 BLOB, " // fastresume
                           "%5 INT DEFAULT -1, " // queue
                           "%6 TEXT," // save path
                           "%7 INT, " // ratio limit
                           "%8 INT, " // seeding time limi
                           "%9 TEXT, " // category
                           "%10 TEXT, " // tags
                           "%11 TEXT, " // name
                           "%12 BOOL, " // has seed status
                           "%13 BOOL, " // temp path disabled
                           "%14 BOOL, " // has root folder
                           "%15 BOOL, " // paused
                           "%16 BOOL, " // forced
                           "%17 BOOL, " // has first-last piece prio
                           "%18 BOOL " // sequential
                           ")"
                           ).arg(TABLE_NAME, COL_HASH, COL_METADATA, COL_FASTRESUME, COL_QUEUE,
                                 COL_SAVE_PATH, COL_RATIO_LIMIT, COL_SEEDING_TIME_LIMIT, COL_CATEGORY)
                            .arg(COL_TAGS, COL_NAME, COL_SEED_STATUS, COL_TMP_PATH_DISABLED, COL_HAS_ROOT_FOLDER,
                                 COL_PAUSED, COL_FORCED, COL_HAS_FIRST_LAST_PIECE_PRIO, COL_SEQUENTIAL));
```

>How are you going to ensure compatibility and control it?

**I am open to ideas.** Currently I use the `user_version` PRAGMA to check that the DB version can be handled by the program.
You have seen my inline comment and responded about backwards compatibility. We can ensure that by refusing to remove columns in newer versions.
If for some serious reason we need to change the format entirely in the DB/table then we do what we have done so far. On 1st run inform the user that there will an incompatibility if he continues with the new version. In this case we update the `user_version` so older clients can have a very easy way to figure out the incompatibility and stop working. And we should go one step further: Require that breaking changes also introduce a "revert" switch, that will rebuild the old format so the user can go back to earlier versions. Like my `Add option to export torrent data to old saving system` commit in this PR.

Your stored procedures idea sounds interesting. I have zero experience with that kind of thing. I don't know the capabilities those can have. And since we don't know what kind of changes we might introduce in the future I don't know if we should rely on that solution. At least in C++ (from new clients) we have the freedom to reconstruct the dbs any way we want and manipulate the data as we see fit.

>How did you decide to migrate from the old saving system?

Do you mean how the code behaves or how I came to the decision to introduce SQLite saving?
Code: Asks the user, parses the fastresumes and populates the relevant table columns. Then deletes the old files.
Decision: Probably explained in opening post. Are you against SQLite?

sledgehammer999 commented:
>The only reliable way to have a consistent torrent queue is to write it all at once. I fixed it recently, but here you degrade again, writing torrent positions separately.

Hmm, I see what you mean. I didn't realize the out-of-sync possibility.  I'll look into it tomorrow to see how to rectify this.

glassez commented:
>he only problem I see is with how I pass the queue number to ResumeDataSavingManager. This doesn't seem necessary and I should remove it.

Yes.
All you need is replacing `Session::saveTorrentsQueue()` content with new code that updates all queue at once.

glassez commented:
About backward compatibility.
>We can ensure that by refusing to remove columns in newer versions.

We must certainly avoid frequent changes in the format of the database. But this should not be an obstacle in the development of the project.
There really isn't much of a problem with that. Because the newer version knows exactly what format the database had in any previous version, it can easily convert the database to the latest format. As your following comment:
>If for some serious reason we need to change the format entirely in the DB/table then we do what we have done so far. On 1st run inform the user that there will an incompatibility if he continues with the new version. In this case we update the user_version so older clients can have a very easy way to figure out the incompatibility and stop working.

>And we should go one step further: Require that breaking changes also introduce a "revert" switch, that will rebuild the old format so the user can go back to earlier versions. 

Do you want to export it to the directly previous format? E.g. if format was changed in v2, v3, and v4, then v4 allow export to v3, v3 to v2, v2 to v1?
>Your stored procedures idea sounds interesting. I have zero experience with that kind of thing. I don't know the capabilities those can have.

I meant the usual CREATE/DROP/ALTER TABLE requests, not the stored procedures. I don't remember exactly, but I don't think SQLite has stored procedures.
>And since we don't know what kind of changes we might introduce in the future I don't know if we should rely on that solution.

You don't understand the idea itself. We should not look to the future, but rather to the past.
When some new version updates the database format, it knows exactly what has been changed and how to get it back. And it just stores this sequence of actions as a set of SQL queries. E.g. if it creates new table in new format then it stores "DROP TABLE" query etc.
Although in fact, things can be much more complicated, and we can't do without more advanced logic (which unfortunately is not supported). Therefore, the priority is to plan the structure as carefully as possible to avoid its frequent modification.

sledgehammer999 commented:
I'll ignore your tone and I'll answer for your linked comments.

If I sum up correctly the 1st comment: DB should be initialized/created by Session. Also you allude that the importing should be done by Session. I disagree with that. Importing torrents is an upgrade step. Did you forget that the previous upgrade code handled the import itself without involving the Session?
However, I suppose I could move into Session the initialization/creation of the DB (via a static class method).
If I understood wrongly your comment, please correct me.

About the 2nd comment: Personally I don't see a pro in **us** saving the native libtorrent fastresume fields as separate columns. I don't see what we can gain from it. **If you have a pro I am happy to hear it.** For libtorrent 1.2: When we tell it to generate a fastresume it still populates/calculates all the fields. It also provides convenience functions to convert to/from add_torrent_params->bencoded fastresume. If we want to go with separate columns, we would still need to write a function to extract the add_torrent_params fields and then populate them (for startup). It sounds to me superfluous. It would make sense if we make direct use of any of the native fields. But we don't IIRC.

For clarity: I started this PR because I thought that the multitude of small files created (current way) hurts performance and possibly results in corruption with big queues. I didn't consider the format/architecture of what we saved. I didn't see it as a problem.

>then v4 allow export to v3, v3 to v2, v2 to v1?

Yes. I think that makes more sense and possibly makes it easier on ourselves to not carry obsolete code for too long.

>I meant the usual CREATE/DROP/ALTER TABLE requests, not the stored procedures. I don't remember exactly, but I don't think SQLite has stored procedures.

OK. But this is probably moot if we have the option to downgrade ourselves in new versions.

>You don't understand the idea itself. 

I only meant that SQL might be limited on what it can do to manipulate the DB and allow us to construct a list of downgrade commands.

>This fragment is incorrect.

You're right. The `if` check must be removed. A previously non-seed torrent that became seed will keep its old queue number with that check and the queue will be out of sync.

PS: If I forgot to answer to something, don't hesitate to directly quote it or link to it. This discussion is big and spans several days with breaks. I can't keep everything in my mind cache.

glassez commented:
>I'll ignore your tone

I'm sorry, I'm only human. Just ignore my emotions, focus on the information.

>This discussion is big and spans several days with breaks. I can't keep everything in my mind cache.

Try my way. Don't remove an appropriate email notification until you answer the comment.
>You're right. The if check must be removed. 

Another way is to store previous queue and compare with the current one and then update only changed torrents. Maybe it will be faster in case of a large number (1000, 10000) of torrents.

>I only meant that SQL might be limited on what it can do to manipulate the DB and allow us to construct a list of downgrade commands.

I suggest we not continue this part of the discussion. It's just a distraction for the brain. I think it is unreachable in general (for example, database format changes can be accompanied by other changes that cannot be undone in this way).

>If I understood wrongly your comment, please correct me.

Yes, probably.
The main idea is not where exactly this "upgrade" code will be located, but it is not to break the basic logic of the application. You can do it wherever you want (within reason), even using a separate program. Your next example demonstrates this well (though you made it for opposite purposes):
>Importing torrents is an upgrade step. Did you forget that the previous upgrade code handled the import itself without involving the Session?

Now remember how we got rid of that code. We just threw it away.
And now try to throw away your current code - you will get an inoperable program. This is your mistake.

glassez commented:
>>then v4 allow export to v3, v3 to v2, v2 to v1?
>
>Yes. I think that makes more sense and possibly makes it easier on ourselves to not carry obsolete code for too long.

It looks a little redundant to me. As I understand it, you should still support exporting to the old saving system for quite some time (at least until v4.2 becomes stable enough and is used by default on all supported platforms/OSes/Distros). So why not just have it only? Each version should be able to upgrade from any previous version of 4.2 and from old saving system but it can allow downgrade to old saving system only.

glassez commented:
As for the "data format" issue, I will no longer insist on my opinion if this comment does not convince you. I leave it to you to make your "suzerain's right" to make the final decision. But do not forget that all those mistakes that could have been avoided, will then have to be corrected. They're on your conscience.

>It also provides convenience functions to convert to/from add_torrent_params->bencoded fastresume. If we want to go with separate columns, we would still need to write a function to extract the add_torrent_params fields and then populate them (for startup).

If you're only afraid of the need to create alternative helpers, then I can help you with that. It doesn't look so complicated to me.
As for the convenience of "fastresume data saving" feature of libtorrent, I have never considered it convenient in terms of qBittorrent design. This can only be convenient for a less integrated app that treats libtorrent as some sort of "black box". We are constantly monitoring the status of each torrent, and this "feature" creates only inconvenience for us.
I'm not sure yet we'll be able to avoid it completely. I'm examining it in detail now.

>Personally I don't see a pro in us saving the native libtorrent fastresume fields as separate columns. I don't see what we can gain from it. If you have a pro I am happy to hear it.

They were in my previous comment. Repeat them:
1. Storing data in separate columns should make it easier to manage database and debug the application. *Even now we require special editor to review bencoded files. If we'll want to see this data from DB we need additionally to save it in some file before open in editor.*
2.  In many cases we can avoid all data generation when we need to update only one field. I.e. when some even occurs we receive an appropriate alert so we can update just one (or several) field instead of asking full data from libtorrent. *This reduces the asynchrony and exposure to data loss in the event of a failure.*
3. Some fields can be reused (e.g. we can get rid from qBt-paused, qBt-forced, qBt-sequential).
4. We can avoid redundant bdecoding (at least with libtorrent-1.2). The fact that it has a "convenient" helper for this does not mean that the decoding process isn't excessive itself. Much simpler (from the performance perspective) to fill add_torrent_params from database table (field-to-field) than to parse the bencoded data. Even in case of libtorrent-1.1, we only need to generate the bencoded data from database table fields when we start up torrents. The speed of data generation (as a rule) is much higher than the parsing speed, the speed difference is measured in orders of magnitude.

No matter what decision you make when you consider this comment, be sure to notify me so we can proceed further based on it.

glassez commented:
@sledgehammer999, I have some ideas for "helper" code. Don't you mind if I add commits here from time to time? Then you can squash them. Or should I paste it in comments instead?

sledgehammer999 commented:
@glassez sorry for the lack of comments. I'll be busy for a couple of days. About the fastresume: I see your point and I am thinking of agreeing with you. Give me a few days to see things.

sledgehammer999 commented on src/base/bittorrent/session.h:
```diff
@@ -468,11 +468,10 @@ namespace BitTorrent
-        void handleTorrentCategoryChanged(TorrentHandle *const torrent, const QString &oldCategory);
-        void handleTorrentTagAdded(TorrentHandle *const torrent, const QString &tag);
-        void handleTorrentTagRemoved(TorrentHandle *const torrent, const QString &tag);
+        void handleTorrentNameChanged(TorrentHandle *const torrent, const QString &torrentName);
```
TorrentHandle::name() doesn't expose the same thing. Nothing does.

glassez replied:
I have commented not only on this particular method, but all similar changes.
As for this particular case, I'd better change the handling of the torrent name inside TorrentHandle class.

sledgehammer999 replied:
Personally, I don't see this as a problem.
Do you want to provide a diff that implements this in the "correct" way?

glassez replied:
>Do you want to provide a diff that implements this in the "correct" way?

Then it will be easier to fix such things in the form of separate "prerequisite" PRs.

sledgehammer999 commented on src/base/bittorrent/session.h:
```diff
@@ -484,11 +483,16 @@ namespace BitTorrent
         void handleTorrentTrackerError(TorrentHandle *const torrent, const QString &trackerUrl);
+        void handleTorrentHasRootChanged(TorrentHandle *const torrent);
+        void handleTorrentSequentialToggled(TorrentHandle *const torrent);
+        void handleTorrentFirstLastPieceToggled(TorrentHandle *const torrent, bool enabled);
```
I do it this way because `TorrentHandle::hasFirstLastPiecePriority()` does a potentially expensive calculation.

glassez replied:
I'd better change its implementation inside TorrentHandle class.

sledgehammer999 replied:
I can't find a way to make it inexpensive

glassez replied:
>I can't find a way to make it inexpensive

Just ask yourself why do we need recalculate this property each time we're getting its value?

glassez commented on src/app/upgrade.cpp:
```diff
@@ -0,0 +1,376 @@
+        libt::entry resumeData(libt::entry::dictionary_t);
+        resumeData = fast;
+
+        resumeData["qBt-savePath"] = query.value(4).toString().toStdString();
```
Nothing inside this method indicates the layout of the query result. It would be better to use field names instead of indexes.

LordNyriox replied:
@glassez:  **Off-Topic**:  The plural form of "index" is "indices".

glassez replied:
Aren't both forms acceptable?
https://grammarist.com/usage/indexes-indices/

LordNyriox replied:
@glassez:  From the same link (emphasis mine): 

>Meanwhile, **indices is generally preferred in mathematical, financial, and technical contexts**, while indexes is relatively common in general usage.

glassez replied:
OK, thanks.

sledgehammer999 replied:
> It would be better to use field names instead of indexes.

Unforunately they don't correlate one-to-one. The index numbers are dependent on what you asked with your `SELECT` statement.

glassez replied:
>The index numbers are dependent on what you asked with your SELECT statement.

Yes. But I can't see what you asked with your SELECT statement since it is exist in somewhere else. But the field names speak for themselves (I mean field/column names from SQL table).

sledgehammer999 replied:
>(I mean field/column names from SQL table).

`QSqlQuery::value(const QString &name)` the docs say that this method is inefficient.

Bear in mind that `prepareTorrentData()` is a helper function called from one place only. So the reader/reviewer can backtrack a little bit. Is this sufficient to not seek out a more rigid solution?
How about if I put comments to indicate which column each number represents?

glassez replied:
>How about if I put comments to indicate which column each number represents?

This is more or less clear as it is. There's another problem. Do the indexes used match the correct columns? For example, you changed the query but forgot to change the indexes.

glassez replied:
>Bear in mind that `prepareTorrentData()` is a helper function called from one place only.

I wonder why you're passing QSqlQuery here. Why not extract all needed data in outer method and pass it here as dictionary?

glassez commented on src/app/upgrade.cpp:
```diff
@@ -0,0 +1,376 @@
+
+        query.exec(QString("CREATE TABLE %1 "
+                           "("
+                           "%2 TEXT PRIMARY KEY NOT NULL, " // hash
```
I wondered for a long time whether there were really good reasons to break the readability of SQL queries like this. I've only found two possible reasons, but none of them seems to me to be strong enough:
1. An attempt to protect yourself from possible typos in queries. But it looks ridiculous, because the probability to make a typo in any other query word is not less than in the names of tables/fields. In addition, they all appear at the first test runs.
2. Ability to easily change table/field names in the future. This is also doubtful... It may look useful only in the early stages, but once it reaches the first release (in fact, once it will be merged in master), it will become very difficult to change the database format, so it will become almost unclaimed.

sledgehammer999 replied:
Do you mean the inline C++ comments? I did it for readability and better reviewing of the appropriate var types chosen (in the SQL context), since it involves many columns. Then it will be used as quick reference each time we want to write a new SQL statement.

If you mean my general approach in other SQL statements to split them: I just split them in their logical parts to increase readability. And when the line becomes too long.

glassez replied:
Do you really think that when I say about "break the readability of SQL queries", I mean either comments or structural splitting that should improve the readability?
I meant the fact that you separate all names from SQL queries.

sledgehammer999 replied:
Oh you mean the QString substitutions? I did it to prevent typos and to ease refactoring of this PR, in case it is decided to use a different name for eg the table, before the PR is merged.

glassez replied:
> I did it to prevent typos and to ease refactoring of this PR, in case it is decided to use a different name for eg the table, before the PR is merged.

Then par. 2 (from my comment) is applicable to it:
>Ability to easily change table/field names in the future. This is also doubtful... It may look useful only in the early stages, but once it reaches the first release (in fact, once it will be merged in master), it will become very difficult to change the database format, so it will become almost unclaimed.

sledgehammer999 commented:
I have rebased and addressed these:
1. Now the Session is responsible for initialize the DB. The upgrade code doesn't manage it directly.
2. Moved 2 alert handlers with the rest of the alert handlers
3. maybe some other minor modifications that I don't remember.

@glassez I also decided to follow your suggestion and save the fastresume into separate columns. That's because I observed we unnecessarily request new fastresume data to be generated when the trackers/webseeds list modified. Now we wiil be a ble to save it directly ourselves.
This change isn't part of the current PR yet.

Minor question: Should we save all fields of the fastresume into the DB uncoditionally? For example the fastresume has these fields that don't seem important to save, and we could generate them in runtime. What do you say?
List:
```c++
// <field name> = <value>
file-format = libtorrent resume file
file-version = 1
libtorrent-version = 1.1.9.0
```

sledgehammer999 commented on src/base/bittorrent/session.h:
```diff
@@ -484,11 +483,16 @@ namespace BitTorrent
         void handleTorrentUrlSeedsRemoved(TorrentHandle *const torrent, const QList<QUrl> &urlSeeds);
-        void handleTorrentResumeDataReady(TorrentHandle *const torrent, const libtorrent::entry &data);
-        void handleTorrentResumeDataFailed(TorrentHandle *const torrent);
+        void handleTorrentResumeDataReady(const libtorrent::save_resume_data_alert* p);
```
Done

sledgehammer999 commented:
>@glassez I also decided to follow your suggestion and save the fastresume into separate columns.

uhm well, I've done some preliminary research about differences between 1.1 and 1.2. I observed that there are fields which are deprecated/removed from 1.2. Also I observed that during the 1.1 lifetime new fields were added.
So I am bit worried that a possible future change in the fastresume spec might leave us in a difficult situation. We might end up using deprecated columns in our DB or not being able to save new columns (important if new columns fix issues like not remembering the last_scrape). We would need to closely monitor libtorrent development and release new version which will create a different table schema.

If we save/load the fastresume itself verbatim, we don't need to worry since libtorrent will handle everything correctly.

Am I worried too much about this?
Maybe some middleground, like directly editing the **stored** fastresume to edit the tracker list instead of requesting libtorrent to generate one?

Chocobo1 commented on src/base/bittorrent/db.h:
```diff
@@ -0,0 +1,58 @@
+{
+    namespace DB
+    {
+        const char MAIN_CONNECTION_NAME[] = "torrentsMain";
```
Minor note, it seems these variables are always converted to QString, so you can store them in QString instead and also initialize them with QStringLiteral macro.

glassez commented:
>the fastresume has these fields that don't seem important to save, and we could generate them in runtime. What do you say?

We don't need to store them.

glassez commented:
Note: all my current and further comments about this PR assume that we use libtorrent-1.2 as the main (and libtorrent-1.1 as compatible).

>I also decided to follow your suggestion and save the fastresume into separate columns. That's because I observed we unnecessarily request new fastresume data to be generated when the trackers/webseeds list modified.

My suggestions are based on the fact that we keep track of most of the torrent data all the time the qBittorrent is running (except for a few not so important "fastresume" fields) so it is redundant to generate the same data again.
>So I am bit worried that a possible future change in the fastresume spec might leave us in a difficult situation...
If we save/load the fastresume itself verbatim, we don't need to worry since libtorrent will handle everything correctly.
Am I worried too much about this?

I don't think that should make us uncomfortable. Unlike previous versions libtorrent-1.2 handle fastresume data not as raw data but as add_torrent_params struct that is a part of libtorrent ABI. Knowing @arvidn's tendency to maintain a compatible ABI throughout the libtorrent branch, we can be sure that it will not change during libtorrent-1.2 life cycle (at least, it will not change often/much).
>Maybe some middleground, like directly editing the **stored** fastresume to edit the tracker list instead of requesting libtorrent to generate one?

In that case, I would suggest another way. Take a closer look. In fact, we have diverse data. Some of them represent "important" properties of the torrent, such as its settings and the current download/upload status (eg., save_path, "file mapping", different kind of priorities, paused, auto_managed etc.; usually this is all that is affected by the user). They are likely to be "stable" for a long time (will not be deleted/changed) so these fields should be stored in separate columns. All remaining data (not so "important";  usually this is all that is generated/obtained by libtorrent itself) can be serialized and stored in a single column (but in this case I would highly recommend using a more readable format than "bencode", e.g. JSON).

arvidn commented:
+1 on this patch. My experience is that iterating a sqlite table is *a lot* faster than listing a directory and loading small files. In my mind, the performance gain alone is sufficient motivation for this.

arvidn commented on src/base/bittorrent/db.h:
```diff
@@ -0,0 +1,58 @@
+        const char FILENAME[] = "torrentData.db";
+        const int VERSION = 1;
+
+        const char TABLE_NAME[] = "torrents";
```
it seems unconventional to define global variables in a header file. if you end up including this in multiple translation units, you'll get multiply defined symbols.

I would say these should either be put in the anonymous namespace and made `cons`, or they should be made `extern` and the values moved into a TU.

victorhooi commented:
As a fan of qBittorrent =) - curious if this is still being worked on?

Is it still slated for qBittorrent 4.2.x?

Any idea of when a test version of qBittorrent with Sqlite support might land, for us to kick the tyres with?

LordNyriox commented:
@sledgehammer999, @glassez:  ??

Ryrynz commented:
Would recommend pushing this to a 4.5 or even a 5.0 Milestone.

Pentaphon commented:
> Would recommend pushing this to a 4.5 or even a 5.0 Milestone.

I agree, I'm not sure why this is in the 4.2.6 milestone.

xavier2k6 commented:
for ref (previous discussion/thoughts) that I came acrosss -> https://github.com/qbittorrent/qBittorrent/issues/7565#issuecomment-355127637 onwards..

txtsd commented:
Why is this continuously being pushed forward? What's the holdup?

glassez commented:
In fact, now there are even more prerequisites to use SQLite (for example, we now often save resume data when changing a particular field, but we still have to generate all the data and overwrite the file).
I would like to implement it in the near future if @sledgehammer999 does not intend to do it himself.

Ryrynz commented:
Any chance the current open issues in 4.3.4 can be moved to their respective 4.4 or 5.0 milestones?
They really shouldn't be pushed out to every minor version milestone release after release.

glassez commented:
I am currently working on optimizing the resume data storage subsystem. The implementation of SQLite database storage support is part of my plans. So I still close this PR for the sake of future implementation.
